crates/node/src/watchdog/mod.rs

```diff
@@ -0,0 +1,277 @@
+                        },
+                        None => {
+                            tracing::error!(
+                                        "Scheduler Health channel error, Health channel closed. Stop health check."
+                                    );
+                            // Alone receiver won't get any message. recv always timeout.
+                            (_, scheduler_health_rx) = tokio::sync::mpsc::channel(1);
+                            new_state.2 = false;
+                        }
+                    }
+                }
+                Err(_) => {
+                    tracing::error!("Scheduler Health signal Timeout, no health signal available.");
+                    timout_counter += 1;
+                    if timout_counter >= gracefull_signal_counter_limit {
+                        new_state.2 = false;
+                    }
+                }
+            }
+        }
+
+        let state_changed = new_state != scheduler_state;
+        scheduler_state = new_state;
+
+        //update watchdog state
+        if state_changed {
+            let mut watchdog_state = watchdog_state.lock().await;
+            if !scheduler_state.0 || !scheduler_state.1 {
+                tracing::warn!(
+                    "watchdog detect critical state loop:{} mempool:{}",
+                    scheduler_state.0,
+                    scheduler_state.1
+                );
+                *watchdog_state = WatchDogState::Critical;
+            } else if !scheduler_state.2 {
+                tracing::warn!("watchdog detect Graceful issue.");
+                *watchdog_state = WatchDogState::Graceful;
+            } else {
+                *watchdog_state = WatchDogState::Alive;
+            }
+        }
+    }
+}
+
+// Start the local server and serve the watch dog requert.
+// WatchDogState::Alive => 200
+// WatchDogState::Graceful => 204
+// WatchDogState::Critical => 503
+async fn serve_files(
+    bind_addr: SocketAddr,
+    watchdog_state: Arc<Mutex<WatchDogState>>,
+) -> Result<()> {
+    let listener = TcpListener::bind(bind_addr).await?;
+
+    tokio::spawn({
+        async move {
+            tracing::info!(
+                "Watchdog listening for http at {}",
+                listener
+                    .local_addr()
+                    .expect("http listener's local address")
+            );
+
+            loop {
+                match listener.accept().await {
+                    Ok((stream, _from)) => {
+                        let io = TokioIo::new(stream);
+                        tokio::task::spawn({
+                            let watchdog_state = watchdog_state.clone();
+                            async move {
+                                if let Err(err) = http1::Builder::new()
+                                    .serve_connection(
+                                        io,
+                                        service_fn(|_| {
```

Comment:
Having specific handler for `/healthz` would be good as there might be need for separate `/ready` or in some cases just a super simple '/' to get dummy HTTP 200 OK if the process exists.